cohortextractor2/query_engines/spark_dialect.py

```diff
@@ -95,6 +98,12 @@ def __init__(self, dialect):
 
 
 class SparkDialect(HiveHTTPDialect):
+    """Customisation of the base Hive dialect.
+
+    It is customised to add support for use with sqlalchemy ORM DDL and
+    connecting connecting to a Databricks hosted Spark.
```

Comment:
*connecting connecting

Reply:
As a naive drive-by reader, I don't understand the docstring here, because I'm wondering why a _sqlalchemy dialect_ needs to be customised for use with the _sqlalchemy ORM_?

Reply:
As it's optionally customised for Databricks, should it be called `SparkOrDatabricksDialect`, or something like that?

Reply:
And perhaps the docstring should explain the switching logic?

Reply:
> As a naive drive-by reader, I don't understand the docstring here, because I'm wondering why a sqlalchemy dialect 
> needs to be customised for use with the sqlalchemy ORM?

Sqlalchemy does a lot more that just SQL, so its dialects provide the nitty gritty support multiple things about a DB, basically:
a) query SQL (which is usually fairly close to ANSI)
b) DDL SQL (which often drifts a bit more from ANSI IME)
c) Metadata inspection commands (usually very DB dependant)
d) Connection set up (which depends on the DBAPI used by the dialect)

The customisations @dave previously added are about type mapping between python and SQL (the date stuff), and pretending to support primary keys (which Spark does not support, but Sqlalchemy's ORM requires). He also handled some differences in error strings in order to raise the right exception for Sqlalchemy ORM's `create_all()` to understand.

The customisations I added were about switching which DBAPI we use, and handling it's the difference in strings output for partition tables.

Ultimately, this is domain knowledge about how Sqlalchemy works. We should probably have a primer like what I've written above, but not in this classes docstring I think.

Reply:
> As it's optionally customised for Databricks, should it be called SparkOrDatabricksDialect, or something like that?

Maybe, but there are there are two reasons its just called SparkDialect at the minute:

1) Databricks **is** Spark. It's SQL is virtually identical to the open source spark. Some of the customisations we've done to the dialect (date stuff, exception wrapping) apply to both the opensource spark we're using **and** databricks (as basically pyhive is a lagging a bit it seems).

2) We currently only support a single dialect hardcoded for a backend, for good reasons. So, for the "Databricks" backend (not its real name), we can only have one dialect. And as that dialect needs to be used to both run tests locally against a container and against a real databricks, it has to support both.

The only other option would be rehauling our backend model to support multiple dialects, or monkeypatching a spark only dialect in when when re run tests.  @dave and I felt that a 'smart' SparkDialect that can switch between connecting to databricks or regular spark (the only significant difference) was the better course of action.

Reply:
I do think the docstring could be re-worded, it's still valid but originally written before the full picture had settled, I will update it and you can see if it makes more sense.

Reply:
I do think the docstring could be re-worded, it's still valid but originally written before the full picture had settled, I will update it and you can see if it makes more sense.